File: src/21. Merge Two Sorted Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

function mergeTwoLists(l1, l2) {
  const dummyHead = new ListNode(undefined)
  let current = dummyHead
  while (l1 && l2) {
    if (l1.val > l2.val) {
      current.next = l2
      l2 = l2.next
    } else {
      current.next = l1
      l1 = l1.next
    }
    current = current.next
  }
  if (l1) current.next = l1
  if (l2) current.next = l2
  return dummyHead.next
}

File: src/21. Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// Recursive
// class Solution {
// public:
//     ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
//         if (list1 == NULL) {
//             return list2;
//         }
//         if (list2 == NULL) {
//             return list1;
//         }
//         if (list1->val <= list2->val) {
//             list1->next = mergeTwoLists(list1->next, list2);
//             return list1;
//         } else {
//             list2->next = mergeTwoLists(list1, list2->next);
//             return list2;
//         }
//     }
// };

// Dummy
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode dummyHead;
        ListNode* current = &dummyHead;
        while (list1 && list2) {
            if (list1->val > list2->val) {
              current->next = list2;
              list2 = list2->next;
            } else {
              current->next = list1;
              list1 = list1->next;
            }
            current = current->next;
        }
        if (list1) current->next = list1;
        if (list2) current->next = list2;
        return dummyHead.next;
    }
};